source/distributed/solution_transfer.cc

```diff
@@ -206,73 +201,83 @@ namespace parallel
         cell_iterator &cell_,
       const typename Triangulation<dim, DoFHandlerType::space_dimension>::
         CellStatus /*status*/,
-      void *data)
+      std::vector<char> &data)
     {
-      typename VectorType::value_type *data_store =
-        reinterpret_cast<typename VectorType::value_type *>(data);
-
       typename DoFHandlerType::cell_iterator cell(*cell_, dof_handler);
 
       const unsigned int dofs_per_cell = cell->get_fe().dofs_per_cell;
-      ::dealii::Vector<typename VectorType::value_type> dofvalues(
-        dofs_per_cell);
-      for (typename std::vector<const VectorType *>::iterator it =
-             input_vectors.begin();
-           it != input_vectors.end();
-           ++it)
+
+      // create buffer for each individual object
+      std::vector<::dealii::Vector<typename VectorType::value_type>> dofvalues(
+        input_vectors.size());
+
+      auto cit_input = input_vectors.cbegin();
+      auto it_output = dofvalues.begin();
+      for (; cit_input != input_vectors.cend(); ++cit_input, ++it_output)
         {
-          cell->get_interpolated_dof_values(*(*it), dofvalues);
-          std::memcpy(data_store,
-                      &dofvalues(0),
-                      sizeof(typename VectorType::value_type) * dofs_per_cell);
-          data_store += dofs_per_cell;
+          it_output->reinit(dofs_per_cell);
+          cell->get_interpolated_dof_values(*(*cit_input), *it_output);
         }
+
+      Utilities::pack(dofvalues, data, /*allow_compression=*/false);
     }
 
 
+
     template <int dim, typename VectorType, typename DoFHandlerType>
     void
     SolutionTransfer<dim, VectorType, DoFHandlerType>::unpack_callback(
       const typename Triangulation<dim, DoFHandlerType::space_dimension>::
         cell_iterator &cell_,
       const typename Triangulation<dim, DoFHandlerType::space_dimension>::
         CellStatus /*status*/,
-      const void *               data,
-      std::vector<VectorType *> &all_out)
+      const boost::iterator_range<std::vector<char>::const_iterator> data_range,
+      std::vector<VectorType *> &                                    all_out)
     {
       typename DoFHandlerType::cell_iterator cell(*cell_, dof_handler);
 
       const unsigned int dofs_per_cell = cell->get_fe().dofs_per_cell;
-      ::dealii::Vector<typename VectorType::value_type> dofvalues(
-        dofs_per_cell);
-      const typename VectorType::value_type *data_store =
-        reinterpret_cast<const typename VectorType::value_type *>(data);
 
-      for (typename std::vector<VectorType *>::iterator it = all_out.begin();
-           it != all_out.end();
-           ++it)
+      std::vector<::dealii::Vector<typename VectorType::value_type>> dofvalues =
+        Utilities::unpack<
+          std::vector<::dealii::Vector<typename VectorType::value_type>>>(
+          data_range.begin(), data_range.end(), /*allow_compression=*/false);
+
+#  ifdef DEBUG
```

Comment:
i don't think you need `ifdef` here, the `for` loop with `Assert` will be ok in release mode.